jupyterhub/auth.py

```diff
@@ -210,6 +209,53 @@ def validate_username(self, username):
         """
     )
 
+    post_auth_hook = Any(config=True,
+        help="""
+        An optional hook function that you can implement to do some
+        bootstrapping work during authentication. For example, loading user account
+        details from an external system.
+        
+        This function is called after the user has passed all authentication checks
+        and is ready to successfully authenticate.
+        
+        This maybe a coroutine.
+
+        Example::
+
+            import os, pwd
+            def my_hook(authenticator, handler, authentication):
+                user_data = pwd.getpwnam(authentication['name'])
+                spawn_data = {
+                    'pw_data': user_data
+                    'gid_list': os.getgrouplist(authentication['name'], user_data.pw_gid)
+                }
+                
+                if authentication['auth_data'] is None:
+                    authentication['auth_data'] = {'spawn_data': spawn_data}
+                else:
+                    authentication['auth_data']['spawn_data'] = spawn_data
+                
+            c.Authenticator.post_auth_hook = my_hook
+
+        """
+    )
+
+    def run_post_auth_hook(self, handler, authentication):
+        """
+        Run the post_auth_hook if defined
+
+        .. versionadded: 1.0
+
+        Args:
+            handler (tornado.web.RequestHandler): the current request handler
+            authentication (dict): User authentication data dictionary. Contains the
+                username ('name'), admin status ('admin'), and auth state dictionary ('auth_state').
+        Returns:
+            None
+        """
+        if self.post_auth_hook is not None:
+            return self.post_auth_hook(self, handler, authentication)
```

Comment:
The docstring (rightly so) allows this to be a coroutine, but the code doesn't. It needs `await maybe_future(self.post...`).

I think it's probably appropriate here to have the hook return the new `authenticated` structure, to make it clear that mutating is appropriate.

Reply:
Hah, you know I'm not good with those! I just copied it from the spawner and renamed things. [Maybe it needs to be fixed there as well?](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py#L840)

Are you thinking we pass in the authentication dict and always replace it with the results to force the user to return it? Or maybe do nothing if it returns None?

Should we pass in a deepcopy'd version? Or should we just send it as it is because it's being returned to make the user pay attention to it?

I'll put together a forced return with the same dict for now.